### ISSUE TITLE
Failed to connect over SSH

### ISSUE THREAD

**Issue by rmGFW** [zed-industries/zed#20854]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

zed ssh://xxx-configured-in-ssh-conf-with-ProxyCommand:/path/to/repo

### Environment

Zed: v0.161.2 (Zed)
OS: macOS 14.5.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<img width="713" alt="image" src="https://github.com/user-attachments/assets/d2db9bfd-c1c6-4038-92bb-aa3c72b7a887">
<img width="260" alt="image" src="https://github.com/user-attachments/assets/8aa77e22-859a-48a9-9fe5-308d3ca3d3c6">

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```

```
</details>

**Comment by rmGFW** [comment]
this is the client side log
2024-11-19T16:35:19.18318+08:00 [INFO] Initializing default prettier with plugins {}
2024-11-19T16:35:24.905971+08:00 [ERROR] failed to establish connection: Timeout detected
2024-11-19T16:35:24.911234+08:00 [ERROR] Failed to open project: Timeout detected
2024-11-19T16:35:26.308558+08:00 [INFO] language server 1: unregistered workspace/DidChangeWatchedFiles capability with id de6d8e03-ea29-486e-ae07-5622e9a65162
---

**Comment by notpeter** [comment]
> `zed ssh://xxx-configured-in-ssh-conf-with-ProxyCommand:/path/to/repo`

I want to see if I can trigger the same behavior.

1. Can you provide your ssh version: `ssh -V`
1. Are there multiple jumps? e.g. `Local->B->C->Remote` or just one: `Local->B->Remote`
1. Do any of the hosts require keyboard input? (password prompt, 2FA code, etc) 
1. Are you using AgentForwarding?
1. Are you using short-lived SSH keys / certificates or related solution (e.g. Teleport)?
1. Can you provide the ssh config including the ProxyCommand (e.g. `ProxyCommand ssh machineA nc %h %p` and the configuration for all host in the chain (e.g. Match, Host `*`, Jump host, host blocks, etc).

Feel free to redact, but the more information you can provide the more likely we are to be able to reproduce your setup and trigger the issue you're having.

**Comment by rmGFW** [comment]
> I have seen this behavior -- does your .bashrc on the remote host do something like
> 
> ```shell
> exec $(which zsh)
> ```
> 
> ?
> 
> If so, can you try updating it to look like this?
> 
> ```shell
> if test -t 1; then
>   exec $(which zsh)
> fi
> ```
> 
> (alternatively, move `exec [some-other-shell]` to your `.bash_profile`)

you are right , i did ,  but after remove exec which zsh ,  timeout error still came out

**Comment by rmGFW** [comment]
r

> > `zed ssh://xxx-configured-in-ssh-conf-with-ProxyCommand:/path/to/repo`
> 
> I want to see if I can trigger the same behavior.
> 
> 1. Can you provide your ssh version: `ssh -V`
> 2. Are there multiple jumps? e.g. `Local->B->C->Remote` or just one: `Local->B->Remote`
> 3. Do any of the hosts require keyboard input? (password prompt, 2FA code, etc)
> 4. Are you using AgentForwarding?
> 5. Are you using short-lived SSH keys / certificates or related solution (e.g. Teleport)?
> 6. Can you provide the ssh config including the ProxyCommand (e.g. `ProxyCommand ssh machineA nc %h %p` and the configuration for all host in the chain (e.g. Match, Host `*`, Jump host, host blocks, etc).
> 
> Feel free to redact, but the more information you can provide the more likely we are to be able to reproduce your setup and trigger the issue you're having.

1. ssh -V output 
ssh -V
OpenSSH_9.6p1, LibreSSL 3.3.6

2. 
<img width="775" alt="image" src="https://github.com/user-attachments/assets/1fdae047-0333-4c80-ac02-6a991d862ee0">

3. no 

4 and 5 , i don't known 

6 . see the picture i have provided in the question 2

thanks! 

by the way , (the vscode remote ssh works well

**Comment by notpeter** [comment]
> Can you provide the ssh config including the ProxyCommand (e.g. ProxyCommand ssh machineA nc %h %p and the configuration for all host in the chain (e.g. Match, Host *, Jump host, host blocks, etc).

I'm particularly interested in your `ProxyCommand` as I believe that is the key to reproducing this.  It seems to begin with an `s`, is it in fact just `ssh some-internal-host`?

Thanks for the rest of the info. I will see if I can reproduce.

**Comment by rmGFW** [comment]
> > you are right , i did , but after remove exec which zsh , timeout error still came out
> 
> Interesting, maybe the behavior of the ssh server makes a difference -- for me, the 'exec' call created issues running non-interactive ssh commands. So simply running something like `ssh [remote] "echo test"` would fail.

ssh [remote] 'echo test' works in my case

**Comment by rmGFW** [comment]
> > Can you provide the ssh config including the ProxyCommand (e.g. ProxyCommand ssh machineA nc %h %p and the configuration for all host in the chain (e.g. Match, Host *, Jump host, host blocks, etc).
> 
> I'm particularly interested in your `ProxyCommand` as I believe that is the key to reproducing this. It seems to begin with an `s`, is it in fact just `ssh some-internal-host`?
> 
> Thanks for the rest of the info. I will see if I can reproduce.

thanks for you reply, the proxy command is a command line tool write by myself , it will create an tunnel over websocket for forwarding.  there is no documentation and close source , but it works well with vscode remote ssh .

**Comment by beniaminzagan** [comment]
@notpeter really seems similar to #25246, in my opinion getting to the bottom of the mentioned issue will probably fix this one as well.

**Comment by notpeter** [comment]
@devzeth Hopefully https://github.com/zed-industries/zed/issues/25246 covers it.
@rmGFW Feel free to check out the propose PR fix for that:
- https://github.com/zed-industries/zed/pull/25314

If you can provide complete steps so I can independently reproduce, please feel free to open a clean issue. But without the necessary bits to reproduce, I'm closing this as unactionable.

Good luck!
